### PR TITLE
[ADD] module for RMA location can be selected on the warehouses

### PR DESCRIPTION
**Pull request by yaniaular**
module for RMA location can be selected on the warehouses

### CODE REVIEW

**Comment by yaniaular** [comment]
We are working on this PR to make it perfect to merge with OCA/rma.
The modules presented here are part of a collection of modules for the operations of RMA improved by Vauxoo. Each module will be presented in separate PRs because the PR #26 was too heavy for review.. Some dependencies needed.

File: crm_rma_stock_location/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+ 'author': "Vauxoo,Camptocamp,Odoo Community Association (OCA)",
+ 'maintainer': 'Camptocamp',
+ 'license': 'AGPL-3',
+ 'category': 'Generic Modules/CRM & SRM',
+ 'depends': ['stock',
+             'procurement',
+             'crm_rma_location_rma',
+             ],
+ 'description': """
+RMA Stock Location
```
Comment (pedrobaeza) [inline crm_rma_stock_location/__openerp__.py]: Extract this on a README.rst file

File: crm_rma_stock_location/product.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+                return
+            for wh in wids:
+                if wh.lot_rma_id:
+                    location_ids.add(wh.lot_rma_id.id)
+            if not location_ids:
+                return
+            ctx['location'] = list(location_ids)
+
+        domain_products = [('product_id', 'in', [self.id])]
+        domain_quant, domain_move_in, domain_move_out = \
```
Comment (pedrobaeza) [inline crm_rma_stock_location/product.py]: Put `self.with_context(ctx)` in a variable and then use it across the method.

File: crm_rma_stock_location/stock.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+#
+#    You should have received a copy of the GNU Affero General Public License
+#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+##############################################################################
+
+from openerp.osv import osv
+
+
+class stock_move(osv.Model):
```
Comment (pedrobaeza) [inline crm_rma_stock_location/stock.py]: CamelCase

Comment (pedrobaeza) [inline crm_rma_stock_location/stock.py]: You can use new API for this little one

File: crm_rma_stock_location/product.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+              'not in',
+              ('done', 'cancel', 'draft'))] + domain_products
+        domain_quant += domain_products
+        if context.get('lot_rma_id'):
+            if context.get('lot_rma_id'):
+                domain_quant.append(('lot_rma_id', '=', context['lot_rma_id']))
+            moves_in = []
+            moves_out = []
+        else:
+            moves_in = self.env['stock.move'].\
```
Comment (pedrobaeza) [inline crm_rma_stock_location/product.py]: Cache in a variable also `self.env['stock.move'].with_context(ctx)`

File: crm_rma_stock_location/product.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+import openerp.addons.decimal_precision as dp
+from openerp.tools.float_utils import float_round
+
+
+class ProductProduct(models.Model):
+
+    _inherit = 'product.product'
+
+    @api.one
+    @api.depends('rma_qty_available', 'rma_virtual_available')
```
Comment (pedrobaeza) [inline crm_rma_stock_location/product.py]: This depends is not correct. You should put the source fields that modifies this computation. Leaving store=False, you don't need to put anything.

File: crm_rma_stock_location/product.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+            float_round(quants.get(self.id, 0.0),
+                        precision_rounding=self.uom_id.rounding)
+
+        self.rma_virtual_available = \
+            float_round(quants.get(self.id, 0.0) +
+                        moves_in.get(self.id, 0.0) -
+                        moves_out.get(self.id, 0.0),
+                        precision_rounding=self.uom_id.rounding)
+
+    rma_qty_available = fields.Float(compute='_rma_product_available',
```
Comment (pedrobaeza) [inline crm_rma_stock_location/product.py]: Maybe without storing, the performance slows down, but you have to put correctly depends (maybe creating auxiliary fields).

File: crm_rma_stock_location/product.py (hunk 2 above)
Comment (nhomar) [inline crm_rma_stock_location/product.py]: 2015-08-05 15:26 GMT-05:00 Pedro M. Baeza notifications@github.com:

> Put self.with_context(ctx) in a variable and then use it across the
> method.

+1

## 

---

Saludos Cordiales

CEO at Vauxoo https://www.vauxoo.com Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
http://about.me/nhomar?promo=email_sig

File: crm_rma_stock_location/product.py (hunk 6 above)
Comment (nhomar) [inline crm_rma_stock_location/product.py]: 2015-08-05 15:32 GMT-05:00 Pedro M. Baeza notifications@github.com:

> Maybe without storing, the performance slows down, but you have to put
> correctly depends (maybe creating auxiliary fields).

@yaniaular, you should extend here the usage of thi field in order to allow
@osvalr make the change.

## 

---

Saludos Cordiales

CEO at Vauxoo https://www.vauxoo.com Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
http://about.me/nhomar?promo=email_sig

File: crm_rma_stock_location/stock.py (hunk 3 above)
Comment (nhomar) [inline crm_rma_stock_location/stock.py]: 2015-08-05 15:29 GMT-05:00 Pedro M. Baeza notifications@github.com:

> You can use new API for this little one

use models.Model please.

## 

---

Saludos Cordiales

CEO at Vauxoo https://www.vauxoo.com Odoo's Gold Partner.

[image: --]
Nhomar Hernandez
[image: http://]about.me/nhomar
http://about.me/nhomar?promo=email_sig

**Comment by osvalr** [comment]
@yaniaular can you make the changes suggested in [here](https://github.com/OCA/rma/pull/37#discussion_r36350395) and [here](https://github.com/OCA/rma/pull/37#discussion_r36350484)

File: crm_rma_stock_location/views/product_template.xml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+<openerp>
```
Comment (yaniaular) [inline crm_rma_stock_location/views/product_template.xml:2]: @osvalr  Add <?xml version="1.0" encoding="utf-8"?>

Comment (osvalr) [inline crm_rma_stock_location/views/product_template.xml:2]: @yaniaular I see the xml header already set

**Comment by osvalr** [comment]
Hi @yaniaular @nhomar @pedrobaeza review ASAP

**Comment by yaniaular** [comment]
@osvalr where is the test for quantity virtual?

**Comment by yaniaular** [comment]
@osvalr Remember take in account the case when the picking is created from claim. Picking in and PIcking out with respect to location rma..

**Comment by osvalr** [comment]
Hi @yaniaular 
I tried to use yml test adding them again to prove if it was okay, but getting this [error](https://travis-ci.org/OCA/rma/jobs/92766905#L262), so do you agree with remove it and just leave unittest instead?

**Comment by osvalr** [comment]
Hi @yaniaular @nhomar ,
Now this module contains all business logic that `crm_rma_advance_location` had in OCA/rma#38, so as you can see this is ready

**Comment by nhomar** [comment]
@osvalr @pedrobaeza @yaniaular I will close this PR in order to make a better PR that have control and information about everything done here, it will not change anything but will allow have a cleanest approval process.

**Comment by osvalr** [comment]
**`UPDATE:`** OCA/rma#67 includes this proposal

**Pull request closed without merging**